src/OpenTK.Compute/OpenCL/CL.cs

```diff
@@ -1,1543 +1,3192 @@
+            [In] uint numberOfDevices,
+            [In] CLDevice[] deviceList,
+            [In] string options,
+            [In] uint numberOfInputDevices,
+            [In] CLProgram[] inputHeaders,
+            [In] string[] headerIncludeNames,
+            [In] IntPtr notificationCallback,
+            [In] IntPtr userData);
+
+        /// <summary>
+        /// Introduced in OpenCL 1.2
+        /// </summary>
+        public static CLResultCode CompileProgram(
+           this CLProgram program,
+            CLDevice[] deviceList,
+            string options,
+            CLProgram[] inputHeaders,
+            string[] headerIncludeNames,
+            ClEventCallback callback,
+            IntPtr userData)
+        {
+            return CompileProgram(program, (uint)(deviceList?.Length ?? 0), deviceList, options, (uint)(inputHeaders?.Length ?? 0), inputHeaders,
+               headerIncludeNames, callback == null ? IntPtr.Zero : Marshal.GetFunctionPointerForDelegate(callback), userData);
+        }
+
+        /// <summary>
+        /// Introduced in OpenCL 1.2
+        /// </summary>
+        [DllImport(LibName, CallingConvention = CallingConvention, EntryPoint = "clLinkProgram")]
+        public static extern CLProgram LinkProgram(
+            [In] this CLContext context,
+            [In] uint numberOfDevices,
+            [In] CLDevice[] deviceList,
+            [In] string options,
+            [In] uint numberOfPrograms,
+            [In] CLProgram[] inputPrograms,
+            [In] IntPtr notificationCallback,
+            [In] IntPtr userData,
+            [Out] out CLResultCode resultCode);
+
+        /// <summary>
+        /// Introduced in OpenCL 1.2
+        /// </summary>
+        public static CLProgram LinkProgram(
+            this CLContext context,
+            CLDevice[] deviceList,
+            string options,
+            CLProgram[] inputPrograms,
+            ClEventCallback callback,
+            IntPtr userData,
+            out CLResultCode resultCode)
+        {
+            return LinkProgram(context, (uint)(deviceList?.Length ?? 0), deviceList, options, (uint)(inputPrograms?.Length ?? 0), inputPrograms,
+               callback == null ? IntPtr.Zero : Marshal.GetFunctionPointerForDelegate(callback), userData, out resultCode);
+        }
+
+        /// <summary>
+        /// Introduced in OpenCL 2.2
+        /// </summary>
+        [DllImport(LibName, CallingConvention = CallingConvention, EntryPoint = "clSetProgramSpecializationConstant")]
+        public static extern CLResultCode SetProgramSpecializationConstant(
+            [In] this CLProgram program,
+            [In] uint specId,
+            [In] nuint specSize,
+            [In] IntPtr specValue);
+
+        /// <summary>
+        /// Introduced in OpenCL 2.2
+        /// </summary>
+        public static unsafe CLResultCode SetProgramSpecializationConstant<T>(
+            this CLProgram program,
+            uint specId,
+            T specValue)
```

Comment:
The original api takes the `specValue` by reference so so should we. This should be either `in T specValue` or `ref T specValue`.

Reply:
Fixed